File: src/services/QueryService.ts
import { OllamaService } from "./OllamaService.js";
import { ChromaService } from "./ChromaService.js";
import { Config } from "../config/Config.js";
import { RedisCacheService } from "../Cache/RedisCacheService.js";

export class QueryService {
  private cache: RedisCacheService;

  constructor(
    private ollamaService: OllamaService,
    private chromaService: ChromaService,
    private config: Config
  ) {
    this.cache = new RedisCacheService();
  }

  async askSimple(question: string): Promise<string> {
    return await this.ollamaService.generate(question);
  }

  async *askStream(history: { role: string; content: string }[], fileId?: string): AsyncGenerator<string> {

    const lastUserMessage = history.filter(m => m.role === 'user').pop();
    if (!lastUserMessage) throw new Error("No user message found");

    const question = lastUserMessage.content.trim();

    const exactCache = await this.cache.getExact(question);
    if (exactCache) {
      console.log("⚡ Exact Cache Hit");
      yield exactCache;
      return;
    }

    const queryEmbedding = await this.ollamaService.embed(question);

    const fuzzyCache = await this.cache.getFuzzy(queryEmbedding);
    if (fuzzyCache) {
      console.log("⚡ Fuzzy Cache Hit");
      yield fuzzyCache;
      return;
    }

    const collection = await this.chromaService.getCollection();

    const queryOptions: {
      queryEmbeddings: number[][];
      nResults: number;
      where?: Record<string, string>;
    } = {
      queryEmbeddings: [queryEmbedding],
      nResults: this.config.retrieval.nResults
    };

    if (fileId !== undefined && fileId !== null && fileId !== '') {
      queryOptions.where = { fileId: fileId };
    }

    const results = await collection.query(queryOptions);

    let context = "";
    if (results.documents?.length > 0 && results.documents[0]?.length > 0) {
      context = results.documents[0].join("\n\n---\n\n");
    }

    const systemPrompt = `
Use the following context to answer the user's question. If the context doesn't contain enough information, you can say so or use your general knowledge, but prioritize the context.
IMPORTANT => Don't use markdown or any other formatting. Always return answer in plain text. 

Context:
${context}
`;

    const messages = [
      { role: 'system', content: systemPrompt },
      ...history
    ];

    let finalAnswer = "";

    console.log("🧠 LLM response streaming...");

    for await (const chunk of this.ollamaService.chatStream(messages)) {
      finalAnswer += chunk;
      yield chunk;
    }

    console.log("📌 Saving to cache (Exact & Fuzzy)...");
    this.cache.setExact(question, finalAnswer);
    this.cache.setFuzzy(question, finalAnswer, queryEmbedding);
  }
}

File: BACKEND/python-service/src/services/query_service.py
import asyncio
import ollama
from typing import AsyncGenerator, Dict, Any, List
from ..services.retrieval_service import RetrievalService

class QueryService:
    def __init__(self, retrieval_service: RetrievalService):
        self.retrieval_service = retrieval_service
        self.ollama_service = ollama.AsyncClient()

    def _format_context_with_citations(self, documents: List[str], metadatas: List[Dict[str, Any]]) -> str:
        if not documents:
            return ""

        formatted_parts = []
        for i, doc in enumerate(documents):
            metadata = metadatas[i]
            citation = ""
            if metadata:
                if "pageNumber" in metadata:
                    citation = f"[Page {metadata['pageNumber']}]"
                elif "pageRange" in metadata:
                    citation = f"[Pages {metadata['pageRange']}]"

            if citation:
                formatted_parts.append(f"{citation}\n{doc}")
            else:
                formatted_parts.append(doc)

        return "\n\n---\n\n".join(formatted_parts)

    async def ask_stream(self, history: List[Dict[str, str]], file_id: str) -> AsyncGenerator[Dict[str, Any], None]:
        last_user_message = next((msg for msg in reversed(history) if msg["role"] == "user"), None)
        if not last_user_message:
            raise ValueError("No user message found in history")

        question = last_user_message["content"]

        retrieval_result = self.retrieval_service.retrieve_text(question, file_id)

        if not retrieval_result.text_chunks:
            yield {"type": "text", "data": "I couldn't find relevant information in the documents."}
            return

        context = self._format_context_with_citations(
            [chunk.text for chunk in retrieval_result.text_chunks],
            [chunk.metadata for chunk in retrieval_result.text_chunks],
        )

        system_prompt = f"""Use the following context to answer the user's question. If the context doesn't contain enough information, you can say so or use your general knowledge, but prioritize the context.
IMPORTANT => Don't use markdown or any other formatting. Always return answer in plain text.

Context:
{context}
"""
        messages = [{"role": "system", "content": system_prompt}] + history

        stream = await self.ollama_service.chat(model='llama2', messages=messages, stream=True)

        for chunk in stream:
            if "content" in chunk["message"]:
                yield {"type": "text", "data": chunk["message"]["content"]}
